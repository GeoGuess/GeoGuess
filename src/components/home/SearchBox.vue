<template>
    <div class="search-box">
        <h2>{{ $tc('Home.placeVisited', nbPlaceVisits) }}</h2>

        <div class="search-box__btns">
            <v-btn
                class="search-box__btns__btn"
                rounded
                color="primary"
                large
                @click="openDialog()"
            >
                {{ $t('DialogRoom.singlePlayer') }}
            </v-btn>

            <v-btn
                class="search-box__btns__btn"
                rounded
                color="secondary"
                large
                @click="openDialog(false)"
            >
                {{ $t('DialogRoom.withFriends') }}
            </v-btn>
            <DialogRoom />
        </div>
        <div class="search-box__mapmenu">
            <v-btn
                text
                class="btn-customs"
                color="primary"
                height="50"
                @click="dialogCustom = !dialogCustom"
            >
                <v-icon>mdi-map-plus</v-icon> Create Map
            </v-btn>
            <DialogCustomMap
                :visibility="dialogCustom"
                @change-visibility="changeDialogCustom"
            />
        </div>
    </div>
</template>
<script>
import DialogCustomMap from '@/components/home/DialogCustomMap';
import DialogRoom from '@/components/dialogroom/DialogRoom';
import { mapActions, mapGetters } from 'vuex';
export default {
    components: {
        DialogRoom,
        DialogCustomMap,
    },
    props: {
        dialogCustomOpen: Boolean,
    },
    data() {
        return {
            dialog: false,
            dialogCustom: this.dialogCustomOpen,
        };
    },
    watch: {
        dialogCustomOpen(v) {
            this.dialogCustom = v;
        },
    },
    computed: {
        ...mapGetters(['nbPlaceVisits']),
    },

    mounted() {
        this.loadHistory();
    },
    methods: {
        ...mapActions(['loadHistory']),
        ...mapActions('settingsStore', ['openDialogRoom']),
        openDialog(isSinglePlayer) {
            this.openDialogRoom(isSinglePlayer);
        },
        changeDialogCustom() {
            this.dialogCustom = !this.dialogCustom;

            this.$router.push(this.dialogCustom ? '/custom' : '/');
        },
    },
};
</script>
<style lang="scss">
.search-box {
    h2 {
        text-align: center;
    }

    .v-input {
        font-size: 1.2rem !important;
    }
    &__btns {
        margin-top: 1.125rem;
        display: flex;
        justify-content: space-around;
        width: calc(100% - 50px);
        &__btn {
            width: 40%;
            padding: 0 5em;
            font-size: 1.1rem;
        }
    }
    &__mapmenu {
        text-align: center;
        margin-top: 3rem;
    }
}
@media (max-width: 410px) {
    .search-box {
        .v-input {
            font-size: 1rem !important;
            width: 95% !important;
            margin: auto !important;
        }
        .search-box__btns {
            margin-top: 0;
            flex-direction: column;
            .v-btn {
                width: 80%;
                margin: 2% auto;
            }
        }
    }
}
</style>
