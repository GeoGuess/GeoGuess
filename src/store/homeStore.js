import axios from '@/plugins/axios';
import { validURL } from '@/utils';
import { getLocateString, isGeoJSONValid } from '../utils';
import * as MutationTypes from './mutation-types';
import i18n from '../lang';
import IndexedDBService from '../plugins/IndexedDBService';
import { GeoMap, GeoMapCustom } from '../models/GeoMap';

export default {
    state: () => ({
        map: new GeoMapCustom(),
        loadingGeoJson: false,
        errorMessage: null,
        listMaps: [],
        listAreas: [],
        customsMaps: [],
        history: [],
        streamerMode: localStorage.getItem('streamerMode') === 'true',
    }),
    mutations: {
        [MutationTypes.HOME_SET_GEOJSON](state, geojson) {
            const map = new GeoMapCustom();
            map.name = state.map.name;
            map.geojson = geojson;
            state.map = map;
        },
        [MutationTypes.HOME_SET_NAME_GEOJSON](state, name) {
            state.map.name = name;
        },
        [MutationTypes.HOME_SET_MAP](state, map) {
            state.map = map;
        },
        [MutationTypes.HOME_SET_LISTS](state, lists) {
            state.listMaps = lists.maps;
            state.listAreas = lists.areas;
        },
        [MutationTypes.HOME_SET_LISTS_CUSTOMMAPS](state, customsMaps) {
            state.customsMaps = customsMaps;
        },
        [MutationTypes.HOME_SET_HISTORY](state, history) {
            state.history = history;
        },
        [MutationTypes.HOME_SET_STREAMER_MODE](state, streamerMode) {
            state.streamerMode = streamerMode;
            localStorage.setItem('streamerMode', streamerMode);
        },

        [MutationTypes.HOME_SET_STATUS_GEOJSON](state, status) {
            state.loadingGeoJson = status;
        },
        [MutationTypes.HOME_SET_GEOJSON_ERROR](state, error) {
            state.errorMessage = error;
        },
    },

    getters: {
        geoJsonString(state) {
           if (!state.map || !state.map.geojson) {
                return '';
            }
            return JSON.stringify(state.map.geojson, null, 2);
        },
        geoJson(state) {
            return state.map.geojson;
        },
        isValidGeoJson(state) {
           if (!state.map || !state.map.geojson) {
                return null;
            }
            return isGeoJSONValid(state.map.geojson);
        },
        maps(state) {
            return state.customsMaps
                .map((map) => Object.assign(new GeoMapCustom(), map))
                .concat(
                    state.listMaps.map((map) =>
                        Object.assign(new GeoMap(), map)
                    )
                );
        },
        areasList(state) {
            return state.listAreas.map((map) => ({
                ...map,
                nameLocate: getLocateString(map, 'name', i18n.locale),
                descriptionLocate: getLocateString(
                    map,
                    'description',
                    i18n.locale
                ),
            }));
        },
        nbPlaceVisits(state) {
            return state.history.reduce(
                (a, { rounds }) => a + rounds.length,
                0
            );
        },
    },

    actions: {
        setStreamerMode({ dispatch, commit }, value) {
            commit(MutationTypes.HOME_SET_STREAMER_MODE, value);
            dispatch(
                'alertStore/setAlert',
                value && {
                    title: 'Home.streamerModeActivate',
                    subtitle: 'Home.streamerModeDetails',
                    color: 'streamerMode',
                    icon: 'mdi-twitch',
                },
                { root: true }
            );
        },
        loadPlaceGeoJSON({ commit, state }, place) {
            if (place != null && place != '') {
                if (state.loadingGeoJson) {
                    return;
                }
                commit(MutationTypes.HOME_SET_STATUS_GEOJSON, true);

                commit(MutationTypes.HOME_SET_GEOJSON, null);

                axios
                    .get(
                        `https://nominatim.openstreetmap.org/search/${encodeURIComponent(
                            place.toLowerCase()
                        )}?format=geojson&limit=1&polygon_geojson=1`
                    )
                    .then((res) => {
                        if (
                            res &&
                            res.status === 200 &&
                            res.data.features.length > 0
                        ) {
                            let feature = res.data.features[0];
                            commit(MutationTypes.HOME_SET_GEOJSON, feature);
                            return;
                        }
                        commit(
                            MutationTypes.HOME_SET_GEOJSON_ERROR,
                            'No Found Location'
                        );
                    })
                    .finally(() => {
                        commit(MutationTypes.HOME_SET_STATUS_GEOJSON, false);
                    });
            }
        },
        async loadGeoJsonFromUrl({ commit }, url) {
            if (validURL(url)) {
                // if gist url get raw
                if (RegExp('^(https?://)?gist.github.com/').test(url)) {
                    let urlSplit = url.split('/');
                    if (
                        urlSplit.length > 3 &&
                        urlSplit[urlSplit.length - 1] !== 'raw'
                    ) {
                        urlSplit[urlSplit.length - 3] =
                            'gist.githubusercontent.com';
                        urlSplit.push('raw');
                        url = urlSplit.join('/');
                    }
                }
                const geojson = await axios
                    .get(url)
                    .then((res) => {
                        if (res.status === 200 && res.data) {
                            if (typeof res.data === 'object') {
                                return res.data;
                            } else {
                                return JSON.parse(res.data);
                            }
                        }
                    })
                    .catch((err) => {
                        // eslint-disable-next-line no-console
                        console.log(err);
                    });

                commit(MutationTypes.HOME_SET_GEOJSON, geojson);
            }
        },
        setMapLoaded({ commit }, map) {
            commit(MutationTypes.HOME_SET_MAP, map);
        },
        setGeoJson({ commit }, geojson) {
            commit(MutationTypes.HOME_SET_GEOJSON, geojson);
        },
        async saveGeoJson({ state, dispatch }) {
            await state.map.save();
            dispatch('getListMapsCustoms');
            dispatch(
                'alertStore/setAlert',
                {
                    title: 'Home.mapSavedAlert.title',
                    subtitle: 'Home.mapSavedAlert.subtitle',
                },
                { root: true }
            );
        },
        setGeoJsonString({ commit }, geojson) {
            let obj = null;
            if (geojson !== '') {
                obj = JSON.parse(geojson);
            }
            commit(MutationTypes.HOME_SET_GEOJSON, obj);
        },
        async getListMaps({ commit }) {
            const data = await axios
                .get(
                    process.env.VUE_APP_LIST_MAPS_JSON_URL ||
                        'https://maps.geoguess.games/maps.json',
                    {
                        cache: {
                            maxAge: 1000,
                        },
                    }
                )
                .then((res) => res.data);

            commit(MutationTypes.HOME_SET_LISTS, data);
        },
        async getListMapsCustoms({ commit }) {
            const customsMap = await Promise.resolve(
                IndexedDBService.loadDb().then(async () => {
                    return await IndexedDBService.getAllMaps();
                })
            );
            commit(MutationTypes.HOME_SET_LISTS_CUSTOMMAPS, customsMap);
        },
        loadHistory({ commit }) {
            const history = localStorage.getItem('history')
                ? JSON.parse(localStorage.getItem('history'))
                : [];
            commit(MutationTypes.HOME_SET_HISTORY, history);
        },
    },
};
