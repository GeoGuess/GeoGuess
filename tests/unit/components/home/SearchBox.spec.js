

import SearchBox from '@/components/home/SearchBox.vue';
import axios from '@/plugins/axios';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import VueAxios from 'vue-axios';
import Vuex from 'vuex';
import homeStore from '../../../../src/store/homeStore';
import appInit from '../../testutils/appInit';

const args = appInit(createLocalVue());
let store;
describe('SearchBox.vue', () => {
    beforeEach(() => {
        store = new Vuex.Store({
            modules: {
                homeStore
            },
        });
    });
    it('render', () => {
        Vue.use(VueAxios, axios);
        const wrapper = shallowMount(SearchBox, { ...args, store });
       

        expect(wrapper).toMatchSnapshot();
    });
});
