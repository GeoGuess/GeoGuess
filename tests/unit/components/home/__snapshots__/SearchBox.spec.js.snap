// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox.vue test search Input 1`] = `
<div class="search-box">
  <h2>0 visited place</h2>
  <div class="search-box__search-bar">
    <v-combobox-stub dark="true" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" appendicon="mdi-magnify" backgroundcolor="secondary" height="50" id="search-input" loaderheight="2" autofocus="true" clearicon="$clear" fullwidth="true" placeholder="Enter city, state or country" rounded="true" type="text" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date &amp;&amp; a.getTime() !== b.getTime()) {
    // If the values are Date, compare them as timestamps
    return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" nodatatext="$vuetify.noDataText" items="" itemcolor="primary" itemdisabled="disabled" itemtext="text" itemvalue="value" menuprops="[object Object]" returnobject="true" allowoverflow="true" filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }" searchinput="" delimiters=""></v-combobox-stub>
    <v-btn-stub color="primary" height="50" tag="button" activeclass="" icon="true" type="button" class="btn-customs">
      <v-icon-stub>mdi-map-plus</v-icon-stub>
    </v-btn-stub>
  </div>
  <dialogcustommap-stub></dialogcustommap-stub>
  <div class="search-box__btns">
    <v-btn-stub color="primary" rounded="true" tag="button" activeclass="" large="true" type="button" class="search-box__btns__btn">
      Single Player
    </v-btn-stub>
    <v-btn-stub color="secondary" rounded="true" tag="button" activeclass="" large="true" type="button" class="search-box__btns__btn">
      With Friends
    </v-btn-stub>
    <dialogroom-stub></dialogroom-stub>
  </div>
</div>
`;
